modin/engines/base/frame/dataframe.py

```diff
@@ -18,12 +18,49 @@
 from pandas.core.dtypes.common import is_numeric_dtype
 from typing import List, Hashable
 
+from modin.engines.modin_dataframe import ModinDataframe
 from modin.backends.pandas.query_compiler import PandasQueryCompiler
 from modin.error_message import ErrorMessage
 from modin.backends.pandas.parsers import find_common_type_cat as find_common_type
 
 
-class BasePandasFrame(object):
+class DimensionError(Exception):
+    """Exception raised if there is a shape mismatch."""
+
+    pass
+
+
+def _check_dtypes(new_df, result_schema):
+    """Check whether the input dataframe conforms to the expected schema.
+
+    Parameters
+    ----------
+        new_df: DataFrame
+            DataFrame whose schema to check.
+        result_schema: dictionary of dtypes
+            Mapping from column labels to data types that represents the types of the output dataframe.
+
+    Raises
+    ------
+        TypeError: If schema of new_df does not match schema expressed by result_schema.
+    """
+    # Should we wrap these dtypes in a call to np.dtype? If not, passing np.int64 to result schema
+    # will fail while np.dtype(np.int64) will not, but if we do, both succeed.
+    # Assumes result_schema is a superset of the columns in new_df
+    new_dtypes = pandas.Series(result_schema)
+    if not (
+        all(new_df.dtypes.isin(new_dtypes).values)
+        and all((new_dtypes[new_df.columns] == new_df.dtypes).values)
+    ):
+        raise TypeError(
+            "Resultant schema:\n{}\nDoes not match expected result schema:\n{}".format(
+                new_df.dtypes, new_dtypes
+            )
+        )
+    return new_df
+
+
+class BasePandasFrame(ModinDataframe):
```

Comment:
Do we want to fully abstract from pandas as base backend? We need to know this to form the right structure/architecture of Modin being discussed in #2957.

Reply:
I think we should distinguish base classes (frame, manager, partitions) and classes for pandas backend. We need to define general API for base classes that has to be overridden in child classes. Those things are currently mixed in base folder. @devin-petersohn , what do you think?

Reply:
I agree @YarShev.